File: killrvideo-service-comments/src/main/java/com/killrvideo/service/comment/grpc/CommentsServiceGrpcValidator.java
package com.killrvideo.service.comment.grpc;

import com.killrvideo.utils.FluentValidator;
import io.grpc.stub.StreamObserver;
import killrvideo.comments.CommentsServiceOuterClass.CommentOnVideoRequest;
import killrvideo.comments.CommentsServiceOuterClass.GetUserCommentsRequest;
import killrvideo.comments.CommentsServiceOuterClass.GetVideoCommentsRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import static org.apache.commons.lang3.StringUtils.isBlank;

/**
 * GRPC Requests Validation Utility class : Implements controls before use request and throw
 * errors if parameter are invalid.
 *
 * @author DataStax Developer Advocates team.
 */
@Component
public class CommentsServiceGrpcValidator {
    private static final Logger LOGGER = LoggerFactory.getLogger(CommentsServiceGrpcValidator.class);

    /**
     * Validate comment On video comment query.
     *
     * @param request        current GRPC Request
     * @param streamObserver response async
     */
    public void validateGrpcRequestCommentOnVideo(CommentOnVideoRequest request, StreamObserver<?> streamObserver) {
        FluentValidator.of("commentOnVideo", request, LOGGER, streamObserver)
                .notEmpty("userId", !request.hasUserId() || isBlank(request.getUserId().getValue()))
                .notEmpty("videoId", !request.hasVideoId() || isBlank(request.getVideoId().getValue()))
                .notEmpty("commentId", !request.hasCommentId() || isBlank(request.getCommentId().getValue()))
                .notEmpty( "comment", isBlank(request.getComment()))
                .validate();
    }

    /**
     * Validate get video comment query.
     *
     * @param request        current GRPC Request
     * @param streamObserver response async
     */
    public void validateGrpcRequestGetVideoComment(GetVideoCommentsRequest request, StreamObserver<?> streamObserver) {
        FluentValidator.of("getVideoComments", request,  LOGGER, streamObserver)
                .notEmpty("video id",
                        !request.hasVideoId() || isBlank(request.getVideoId().getValue()))
                .positive("page size", request.getPageSize() <= 0)
                .validate();
    }

    /**
     * Validate get user comment query.
     *
     * @param request        current GRPC Request
     * @param streamObserver response async
     */
    public void validateGrpcRequest_GetUserComments(GetUserCommentsRequest request, StreamObserver<?> streamObserver) {
        FluentValidator.of("getUserComments", request, LOGGER, streamObserver)
                .notEmpty("userId", !request.hasUserId() || isBlank(request.getUserId().getValue()))
                .positive("page size", request.getPageSize() <= 0)
                .validate();
    }
}

File: killrvideo-service-comments/src/main/kotlin/com/killrvideo/service/comment/grpc/CommentsServiceGrpcValidator.kt
package com.killrvideo.service.comment.grpc

import com.killrvideo.utils.FluentValidator
import killrvideo.comments.CommentsServiceOuterClass.*
import mu.KotlinLogging
import org.apache.commons.lang3.StringUtils
import org.springframework.stereotype.Component

/**
 * GRPC Requests Validation Utility class : Implements controls before use request and throw
 * errors if parameter are invalid.
 *
 * @author DataStax Developer Advocates team.
 */
@Component
class CommentsServiceGrpcValidator {
    private val logger = KotlinLogging.logger {  }
    /**
     * Validate comment On video comment query.
     *
     * @param request        current GRPC Request
     */
    fun validateGrpcRequestCommentOnVideo(request: CommentOnVideoRequest) =
        FluentValidator.of("commentOnVideo", request, logger)
            .notEmpty("userId", !request.hasUserId() || StringUtils.isBlank(request.userId.value))
            .notEmpty("videoId", !request.hasVideoId() || StringUtils.isBlank(request.videoId.value))
            .notEmpty("commentId", !request.hasCommentId() || StringUtils.isBlank(request.commentId.value))
            .notEmpty("comment", StringUtils.isBlank(request.comment))
            .validate()

    /**
     * Validate get video comment query.
     *
     * @param request        current GRPC Request
     */
    fun validateGrpcRequestGetVideoComment(request: GetVideoCommentsRequest) =
        FluentValidator.of("getVideoComments", request, logger)
            .notEmpty("video id", !request.hasVideoId() || StringUtils.isBlank(request.videoId.value))
            .positive("page size", request.pageSize <= 0)
            .validate()

    /**
     * Validate get user comment query.
     *
     * @param request        current GRPC Request
     */
    fun validateGrpcRequest_GetUserComments(request: GetUserCommentsRequest) =
        FluentValidator.of("getUserComments", request, logger)
            .notEmpty("userId", !request.hasUserId() || StringUtils.isBlank(request.userId.value))
            .positive("page size", request.pageSize <= 0)
            .validate()
}
